### ISSUE TITLE
Conda packaging

### ISSUE THREAD

**Issue by bmcfee** [librosa/librosa#202]
Providing a conda distribution might save some headaches at install time, especially when it comes to codec dependencies.

**Comment by bmcfee** [comment]
If we end up making a binstar channel, it would be helpful to include libsamplerate as well.

**Comment by carlthome** [comment]
:+1:

**Comment by stefan-balke** [comment]
hey @bmcfee,

I started to play around with anaconda and the packaging.
Here is a first rough starter for libsamplerate.
The plan is to use this to offer the binaries for all platforms at binstar.

https://github.com/stefan-balke/conda-libsamplerate

I still had some compilation issues there so I forked the repo to make a single change.
Now it seems to build...

**Comment by stefan-balke** [comment]
BTW: Can we solve the issue with scikits.samplerate?

I filed an issue here such that the pypi version gets updated.
https://github.com/cournape/samplerate

**Comment by bmcfee** [comment]
That issue is already fixed in the fork; it just needs to be merged into the main distribution.

**Comment by stefan-balke** [comment]
Let's see if David is moving, otherwise we could also file a new version on pypi, couldn't we?

**Comment by stefan-balke** [comment]
So Erik accepted my **MAJOR** :) PR
Conda should build it now with this recipe.

@bmcfee should we start a channel on binstar?
I think it needs to be an organization and then you can upload packages to it.

**Comment by bmcfee** [comment]
> @bmcfee should we start a channel on binstar?
> I think it needs to be an organization and then you can upload packages to it.

Yeah, that makes sense.  Maybe we should punt this til after #240, by which i mean, finally do #240?  That sounds like a fine post-ismir-deadline project to me.

**Comment by stefan-balke** [comment]
Yeah, sounds good.

Any opinions on what packages should be be covered by the channel?
* librosa
* libsamplerate might be the most important since it made the most trouble so far

What else?

**Comment by bmcfee** [comment]
* scikits.samplerate
* libsndfile and/or libav/ffmpeg

**Comment by stefan-balke** [comment]
Seen that?
https://github.com/bastibe/libsndfile-binaries

It's also here: https://github.com/erikd/libsndfile

I can try to get the builds running and write a conda recipe.

Any opinion about those recipes? Should they be as basic as possible?

And, where should they live? 
https://github.com/conda/conda-recipes ?

**Comment by stefan-balke** [comment]
> And, where should they live? 

Looked up other libs - they usually have a repo for all the needed recipes.
Clone that and then build the binaries on your machine and then upload it to
binstar (anaconda cloud).

**Comment by bmcfee** [comment]
> they usually have a repo for all the needed recipes.

yup, sounds good.

**Comment by bmcfee** [comment]
@stefan-balke have you seen [conda-forge](https://conda-forge.github.io/)?

**Comment by stefan-balke** [comment]
@bmcfee nope, but that looks totally awesome.

I will first try with libsamplerate. Keep you in the loop.

**Comment by stefan-balke** [comment]
Started it here:
https://github.com/conda-forge/staged-recipes/pull/229

**Comment by stefan-balke** [comment]
This one is next:

https://github.com/conda-forge/staged-recipes/pull/231

People start to use python more and more in our lab. Most of them in a Windows-Anaconda environment.

**Comment by stefan-balke** [comment]
This certainly is hell :)

 * libsamplerate: https://github.com/conda-forge/staged-recipes/pull/229
  - windows build has issues...
* libsndfile: https://github.com/conda-forge/staged-recipes/pull/231
  - lots of dependencies missing:
   - Autogen https://github.com/conda-forge/staged-recipes/pull/232
   - Guile https://github.com/conda-forge/staged-recipes/pull/232
   - libffi
   - libunistring
   - bdw-gc

Feel free to jump in here.

**Comment by bmcfee** [comment]
Yikes!  I didn't realize it would be such a pain to get things packaged (and working on windows).

A few thoughts here:

- We don't need libsndfile per se, just one of the codec backends supported by audioread.  ffmpeg would also work.. do you think that would be easier?  What about gstreamer?
- libsamplerate: it's been on my to-do to reimplement the sinc interpolation resampler, either as a standalone python package, or as a PR to scipy.  Maybe it's time we just buckled down and did this?  It wouldn't solve our packing problems immediately, but I like it better as a long-term solution.  What do you think?

**Comment by stefan-balke** [comment]
Hey,

so libsamplerate should be fine, although implementing the interpolator would be a nice thing to understand it better.

ffmpeg has similar dependencies so doing autogen as recipes is not worthless.
The maintainers already said that in some edge cases, they bundle their packages with a pre-compiled binary...libsndfile might be such a case for windows.

**Comment by stefan-balke** [comment]
Alright...I resolved all the dependencies and did lots of PRs.

They will hopefully subsequently merge the stuff to their repos and then let's see.

Maybe I will ask @erikd how-to build libsndfile with Windows and will then update the build recipe for the forge.

**Comment by bmcfee** [comment]
> Alright...I resolved all the dependencies and did lots of PRs.

Awesome!  Thanks :100: for doing this!



> although implementing the interpolator would be a nice thing

... I went ahead and did it yesterday.  Still working out the kinks and fiddling parameters, but with numba, it seems comparably fast to libsamplerate.  Converting to cython (for slightly easier pip distribution) ought to be about as fast as well.  I think it should be totally possible to get out a standalone resample package ahead of librosa 0.5.

**Comment by stefan-balke** [comment]
>  I think it should be totally possible to get out a standalone resample package ahead of librosa 

Awesome! :+1: 

So I will concentrate to get libsndfile or ffmpeg or gstreamer up and running. :)

**Comment by bmcfee** [comment]
@stefan-balke we now have a proper librosa organization -- what do you think about moving the conda-forge recipe over there?

No rush on getting it finalized, just taking a moment to get all our ducks in a row.

**Comment by stefan-balke** [comment]
> what do you think about moving the conda-forge recipe over there?

Yeah, sure. I'll prepare a recipe since libsamplerate is obsolete now 👍

**Comment by bmcfee** [comment]
@stefan-balke have you gotten anywhere on this?  If you like, I can make us a librosa-conda repo under the org for this to live in.

**Comment by stefan-balke** [comment]
Hey,

didn't invest too much time yet.
Normally you would fork the conda-forge repo and do a PR into their recipes repository.

Maybe this is something which needs a deadline :)
Same for finishing up DTW. I'll cover both until ISMIR - word!

**Comment by bmcfee** [comment]
Ah, I see. Well let's fork cf under the librosa org then. I'd like to get this done at least a week before ISMIR so that we can use it in the tutorial at hamr. Do you think that's feasible?

**Comment by stefan-balke** [comment]
Yep, I think so. Only "problematic" dependency is `audioread` but there is a feedstock for `ffmpeg` https://github.com/conda-forge/ffmpeg-feedstock.

**Comment by stefan-balke** [comment]
Hey,

I just took a stab on it and we are in pretty good shape already. I did the recipes for

* audioread
* resampy
* librosa

If you fork I can add them to the repo and we can request the PRs for conda-forge.

**Comment by bmcfee** [comment]
@stefan-balke I'm building conda-forge feedstocks for everything now.  (Sitting down next to the conda-forge devs to work this out made things much simpler. :grin:)

Resampy is in, audioread is done and waiting for merge, and then librosa will be easy.

Do you want me to add you as a maintainer on any of these?

**Comment by bmcfee** [comment]
Done.  https://github.com/conda-forge/librosa-feedstock

**Comment by stefan-balke** [comment]
Awesome.

> Do you want me to add you as a maintainer on any of these?

No, not necessarily.
